crates/types/src/block_sync.rs

```diff
@@ -0,0 +1,380 @@
+use std::{
+    collections::VecDeque,
+    fmt::Debug,
+    ops::RangeInclusive,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, RwLock
+    },
+    task::Waker
+};
+
+use dashmap::DashMap;
+
+/// Producer to block syncing events
+pub trait BlockSyncProducer: Debug + Clone + Send + Sync + Unpin + 'static {
+    /// when a new block is produced. starts new block process
+    /// always assume truthful values are passed
+    fn new_block(&self, block_number: u64);
+
+    /// when a reorg happens, starts syncing process
+    /// always assume truthful values are passed
+    fn reorg(&self, reorg_range: RangeInclusive<u64>);
+}
+
+/// Consumer to block sync producer
+pub trait BlockSyncConsumer: Debug + Clone + Send + Sync + Unpin + 'static {
+    fn sign_off_reorg(
+        &self,
+        module: &'static str,
+        block_range: RangeInclusive<u64>,
+        waker: Option<Waker>
+    );
+    fn sign_off_on_block(&self, module: &'static str, block_number: u64, waker: Option<Waker>);
+    fn current_block_number(&self) -> u64;
+    fn has_proposal(&self) -> bool;
+    fn fetch_current_proposal(&self) -> Option<GlobalBlockState>;
+    fn register(&self, module_name: &'static str);
+
+    fn can_operate(&self) -> bool {
+        !self.has_proposal()
+    }
+}
+
+/// The global block sync is a global syncing mechanism.
+///
+/// every module that interacts with blocks and or other modules that are block
+/// sensitive are registered with the global block sync. The global block number
+/// will only increase when all modules that are registered mark the global sync
+/// as ready to increment. Unlike progressing. Any one module is able to abort
+/// the process (aborts due to reorgs)
+#[derive(Debug, Clone)]
+pub struct GlobalBlockSync {
+    /// state that we are waiting on all sign offs for
+    pending_state:      Arc<RwLock<VecDeque<GlobalBlockState>>>,
+    /// the block number
+    block_number:       Arc<AtomicU64>,
+    /// the modules with there current sign off state for the transition of
+    /// pending state -> cur state
+    registered_modules: DashMap<&'static str, SignOffState>
```

Comment:
Given that we will panic in both `sign_off_on_reorg()` and `sign_off_on_block()` if the incoming state we're claiming to sign off on doesn't match the latest entry in `pending_state` this could be simplified into `DashMap<&'static str, Option<Waker>>`.

There's no reason to store a whole `SignOffState` when we're constantly ensuring that it will only ever be equivalent to the `pending_state` that's on deck.

Reply:
Actually given that `SignOffState` gives me the option to say "I've checked in but I don't have a waker you need to work with" it might be more like `Option<Option<Waker>>` to allow for `Some(None)` from modules that have checked in but don't have a waker to share

Reply:
I agree that we shouldn't store the  block / reorg range. However I am going to keep SignOffState with the two variants rather than flatten two a Option<Option<_>> as it is more reasonable and less obfuscated. Readability is important here imo and theres not really any tradeoff